trecs/models/recommender.py

```diff
@@ -337,22 +348,69 @@ def generate_recommendations(self, k=1, item_indices=None):
         else:
             # scores are U x I; we can use argpartition to take the top k scores
             negated_scores = -1 * s_filtered  # negate scores so indices go from highest to lowest
-            top_k = negated_scores.argpartition(k - 1)[:, :k]
+            # break ties using a random score component
+            scores_tiebreak = np.zeros(
+                negated_scores.shape, dtype=[("score", "f8"), ("random", "f8")]
+            )
+            scores_tiebreak["score"] = negated_scores
+            scores_tiebreak["random"] = self.random_state.random(negated_scores.shape)
+            top_k = scores_tiebreak.argpartition(k - 1, order=["score", "random"])[:, :k]
             # now we sort within the top k
             row = np.repeat(self.users.user_vector, k).reshape((self.num_users, -1))
             # again, indices should go from highest to lowest
-            sort_top_k = negated_scores[row, top_k].argsort()
+            sort_top_k = scores_tiebreak[row, top_k].argsort(order=["score", "random"])
             rec = item_indices[
                 row, top_k[row, sort_top_k]
             ]  # extract items such that rows go from highest scored to lowest-scored of top-k
             if self.is_verbose():
-                self.log(f"Row:\n{str(row)}")
                 self.log(f"Item indices:\n{str(item_indices)}")
                 self.log(
-                    f"Top-k items ordered by preference (low to high) for each user:\n{str(rec)}"
+                    f"Top-k items ordered by preference (high to low) for each user:\n{str(rec)}"
                 )
             return rec
 
+    def choose_interleaved_items(self, k, item_indices):
+        """
+        Chooses k items out of the item set to "interleave" into
```

Comment:
Since there has been significant confusing around what this means, it might be helpful to provide an example of what interleaving means in the docstring. Not totally sure what the best phrasing would be but something like "item set to "interleave" (i.e., insert into specific indices in the recommendation set irrespective of the predicted score)

Reply:
Added a docstring! Let me know if you have suggested modifications!